src/helpers/validators/validateString/index.ts

```diff
@@ -0,0 +1,121 @@
+
+/**
+ * Determines whether a given input string is considered valid based on
+ *   the provided requirements.
+ * @author Austen Money
+ * @param input input string
+ * @param reqs requirements that the provided string should conform to
+ * @returns whether input is considered valid according to reqs - if
+ *   valid, returns a cleaned version of input; if invalid, returns
+ *   a string containing error messages describing which requirements
+ *   were not met.
+ */
+const validateString = (
+  input: string,
+  reqs: StringValidationRequirements,
+): ValidationResult => {
+  // stores all invalid input errors
+  const errorMessages: Array<string> = [];
+  // contains version of input that will be returned
+  let cleanedValue: string = input;
+
+  // remove whitespace if required
+  if (reqs.ignoreWhitespace) {
+    cleanedValue = input.replace(/\s+/g, '');
+  }
+
+  // apply max char requirement
+  if (reqs.minLen) {
+    if (cleanedValue.length < reqs.minLen) {
+      errorMessages.push(`Input must not be under ${reqs.minLen} character(s).`);
+    }
+  }
+
+  // apply max char requirement
+  if (reqs.maxLen) {
+    if (cleanedValue.length > reqs.maxLen) {
+      errorMessages.push(`Input must not be over ${reqs.maxLen} character(s).`);
+    }
+  }
+
+  // apply alphabetical requirement
+  if (reqs.lettersOnly) {
+    for (let i = 0, len = cleanedValue.length; i < len; i++) {
+      const curr = cleanedValue.charCodeAt(i);
+
+      // check that char is an upper or lower-case letter
+      if (!(curr > 64 && curr < 91) && 
+          !(curr > 96 && curr < 123)) {
+        errorMessages.push('Input must not contain non-letters.');
+        break;
+      }
+    }
+  }
+
+  // apply numerical requirement
+  if (reqs.numbersOnly) {
+    for (let i = 0, len = cleanedValue.length; i < len; i++) {
+      const curr = cleanedValue.charCodeAt(i);
+
+      // check that char is a digit
+      if (!(curr > 47 && curr < 58)) {
+        errorMessages.push('Input must not contain non-numbers.');
+        break;
+      }
+    }
+  }
+
+  // apply regex requirement
+  if (reqs.regexTest) {
+    const regex = new RegExp(reqs.regexTest);
+    let result: ValidationResult;
+
+    // validate and create customized error message if description is provided
+    reqs.regexDescription
+      ? result = validateRegex(
+        {
+          input: cleanedValue,
+          regex,
+          regexDescription: reqs.regexDescription,
+        },
+      )
+      : result = validateRegex(
+        {
+          input: cleanedValue,
+          regex,
+        },
+      );
+
+    // if string did not pass regex validation, add error message
+    if (result.isValid === false) {
+      errorMessages.push(result.errorMessage);
+    }
+  }
+
+  // combine all error messages into one string to return
+  let errorMessage = 'The following error(s) occurred:';
+  for (let i = 0; i < errorMessages.length; i++) {
+    errorMessage = `${errorMessage} ${errorMessages[i]}`;
```

Comment:
Would prefer for a newline and tab at the beginning of these:

```suggestion
    errorMessage = `${errorMessage}\n\t${errorMessages[i]}`;
```

preferably, you could use [join](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join):

```ts
  const errorMessage = `The following error(s) occurred:\n\t${errorMessages.join('\n\t')}`;
```

or it can be done with a [forEach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach):

```ts
  let errorMessage = 'The following error(s) occurred:'
  errorMessages.forEach((errorMsg) => {
    errorMessage = `${errorMessage}\n\t${errorMsg}`;
  });
```

and don't tell Gabe, but you can also use [reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce):

```ts
const errorMessage = errorMessages.reduce(
  (allErrMsg, currErrMsg) => {
    return `${allErrMsg}\n\t${currErrMsg}`;
  },
  'The following error(s) occurred:',
);
```

Reply:
The join function is great! Just switched over to using that. I also originally had newline + tab formatting, but talked with Gabe and we determined that formatting will usually be added later on, so having it here might not be helpful. Open to more ideas on that, though!

Reply:
I won’t tell Gabe either! 🤫

I love the idea of using join. Perhaps a comma separated list instead of new lines?